File: qt/gen_qtimer.cpp
// +build ignore

#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QTimer>
#include <QTimerEvent>
#include <qtimer.h>
#include "gen_qtimer.h"
class MiqtVirtualQTimer : public virtual QTimer {
public:
MiqtVirtualQTimer(): QTimer() {};
MiqtVirtualQTimer(QObject* parent): QTimer(parent) {};

virtual ~MiqtVirtualQTimer() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QTimer::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QTimer_MetaObject(const_cast<MiqtVirtualQTimer*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QTimer::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QTimer::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QTimer_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QTimer::qt_metacast(param1);

	}
};
QTimer* QTimer_new() {
return new MiqtVirtualQTimer();
}
QTimer* QTimer_new2(QObject* parent) {
return new MiqtVirtualQTimer(parent);
}
void QTimer_virtbase(QTimer* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QTimer_MetaObject(const QTimer* self) {
	return (QMetaObject*) self->metaObject();
}
void* QTimer_Metacast(QTimer* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QTimer_Tr(const char* s) {
	QString _ret = QTimer::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
bool QTimer_IsActive(const QTimer* self) {
	return self->isActive();
}
int QTimer_TimerId(const QTimer* self) {
	return self->timerId();
}
int QTimer_Id(const QTimer* self) {
	Qt::TimerId _ret = self->id();
	return static_cast<int>(_ret);
}
void QTimer_SetInterval(QTimer* self, int msec) {
	self->setInterval(static_cast<int>(msec));
}
int QTimer_Interval(const QTimer* self) {
	return self->interval();
}
int QTimer_RemainingTime(const QTimer* self) {
	return self->remainingTime();
}
void QTimer_SetTimerType(QTimer* self, int atype) {
	self->setTimerType(static_cast<Qt::TimerType>(atype));
}
int QTimer_TimerType(const QTimer* self) {
	Qt::TimerType _ret = self->timerType();
	return static_cast<int>(_ret);
}
void QTimer_SetSingleShot(QTimer* self, bool singleShot) {
	self->setSingleShot(singleShot);
}
bool QTimer_IsSingleShot(const QTimer* self) {
	return self->isSingleShot();
}
void QTimer_Start(QTimer* self, int msec) {
	self->start(static_cast<int>(msec));
}
void QTimer_Start2(QTimer* self) {
	self->start();
}
void QTimer_Stop(QTimer* self) {
	self->stop();
}
struct miqt_string QTimer_Tr2(const char* s, const char* c) {
	QString _ret = QTimer::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QTimer_Tr3(const char* s, const char* c, int n) {
	QString _ret = QTimer::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QTimer_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQTimer*>( (QTimer*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QTimer_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQTimer*)(self) )->virtualbase_MetaObject();
}
void QTimer_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQTimer*>( (QTimer*)(self) )->handle__Metacast = slot;
}
void* QTimer_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQTimer*)(self) )->virtualbase_Metacast(param1);
}
void QTimer_Delete(QTimer* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQTimer*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qtimer.go
package qt

import (
	"unsafe"
)

type QTimer struct {
	h          uintptr
	isSubclass bool
}

// NewQTimer constructs a new QTimer object.
func NewQTimer() *QTimer {
	ret := newQTimer(QTimer_new())
	ret.isSubclass = true
	return ret
}

// NewQTimer2 constructs a new QTimer object.
func NewQTimer2(parent *QObject) *QTimer {
	ret := newQTimer(QTimer_new2(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QTimer) MetaObject() *QMetaObject {
	return newQMetaObject(QTimer_MetaObject(this.h))
}

func (this *QTimer) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QTimer_Metacast(this.h, param1_Cstring))
}

func QTimer_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QTimer_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QTimer) IsActive() bool {
	return (bool)(QTimer_IsActive(this.h))
}

func (this *QTimer) TimerId() int {
	return (int)(QTimer_TimerId(this.h))
}

func (this *QTimer) Id() TimerId {
	return (TimerId)(QTimer_Id(this.h))
}

func (this *QTimer) SetInterval(msec int) {
	QTimer_SetInterval(this.h, (int)(msec))
}

func (this *QTimer) Interval() int {
	return (int)(QTimer_Interval(this.h))
}

func (this *QTimer) RemainingTime() int {
	return (int)(QTimer_RemainingTime(this.h))
}

func (this *QTimer) SetTimerType(atype TimerType) {
	QTimer_SetTimerType(this.h, (int)(atype))
}

func (this *QTimer) TimerType() TimerType {
	return (TimerType)(QTimer_TimerType(this.h))
}

func (this *QTimer) SetSingleShot(singleShot bool) {
	QTimer_SetSingleShot(this.h, (bool)(singleShot))
}

func (this *QTimer) IsSingleShot() bool {
	return (bool)(QTimer_IsSingleShot(this.h))
}

func (this *QTimer) Start(msec int) {
	QTimer_Start(this.h, (int)(msec))
}

func (this *QTimer) Start2() {
	QTimer_Start2(this.h)
}

func (this *QTimer) Stop() {
	QTimer_Stop(this.h)
}

func QTimer_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QTimer_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QTimer_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QTimer_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QTimer) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QTimer_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QTimer) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QTimer_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QTimer_MetaObject
func miqt_exec_callback_QTimer_MetaObject(self QTimer, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QTimer{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QTimer) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QTimer_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QTimer) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QTimer_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QTimer_Metacast
func miqt_exec_callback_QTimer_Metacast(self QTimer, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QTimer{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
